src/python/zquantum/qcbm/target_thermal_states.py

```diff
@@ -0,0 +1,235 @@
+    """Generates samples from the original target distribution in the form of a list of ising vectors
+    Args:
+        n_samples: the number of samples from the original distribution
+        n_spins: positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+    Returns:
+       samples: array of sample n_spin ising vectors
+    """
+    cumulative = cumulate(n_spins, temperature, hamiltonian_parameters)
+    samples = np.zeros((n_samples, n_spins))
+    y = np.zeros(n_samples)
+    for i in range(n_samples):
+        y[i] = np.random.uniform(low=0, high=1)
+    for i in range(len(y)):
+        idx = bisect.bisect_right(cumulative, y[i])
+        ising_bitstring = int2ising(idx, n_spins)
+        samples[i, :] = ising_bitstring
+    return np.asarray(samples)
+
+
+
+def get_thermal_sampled_distribution(
+    n_samples: int,
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
+) -> typing.Dict:
+    """Generates thermal states sample distribution
+    Args:
+        n_samples: the number of samples from the original distribution
+        n_spins: positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+    Returns:
+       histogram_samples: keys are binary string representations and values are corresponding probabilities.
+    """
+    sample_distribution_dict = {}
+    samples = sample(n_samples, n_spins, temperature, hamiltonian_parameters)
+    histogram_samples = np.zeros(2 ** n_spins)
+    pos_spins_list = []
+    for s in samples:
+        idx = ising2int(s)
+        histogram_samples[idx] += 1.0 / n_samples
+        pos_spins = 0 
+        for elem in s: 
+            if elem == 1.0: 
+                pos_spins += 1 
+        pos_spins_list.insert(idx,pos_spins)
+
+    for spin in range(int(2 ** n_spins)):
+        binary_bitstring = convert_tuples_to_bitstrings([dec2bin(spin, n_spins)])[-1]
+        sample_distribution_dict[binary_bitstring] = histogram_samples[spin]
+
+    return sample_distribution_dict, pos_spins_list
+
+
+
+
+
+def get_sampled_bitstring_distribution(
+    n_samples: int,
+    n_spins: int,
+    temperature: float,
+    hamiltonian_parameters: typing.List[np.array],  # TODO: add docstring
+) -> BitstringDistribution:
+    """Generates thermal states sample distribution as BitstringDistribution object
+    Args:
+        n_samples: the number of samples from the original distribution
+        n_spins : positive number of spins in the Ising system
+        temperature: temperature factor in the boltzman distribution
+    Returns:
+        Sampled Bitstring Distribution.
+    """
+    probabilities = get_thermal_sampled_distribution(
+        n_samples, n_spins, temperature, hamiltonian_parameters
+    )[0]
+    spin_mapping = get_thermal_sampled_distribution(
+        n_samples, n_spins, temperature, hamiltonian_parameters
+    )[1]
```

Comment:
This is taking two calls to `get_thermal_sampled_distribution` which seems unnecessary. You should be able to do:
```
probabilities, spin_mapping = get_thermal_sampled_distribution(n_samples, n_spins, temperature, hamiltonian_parameters)
```

Reply:
Taken out anyway